lib/asn1.ml

```diff
@@ -478,23 +480,21 @@ module Algorithm_identifier = struct
       | Rsa
       | Ec
       | Dh
-      | Unknown of Asn.OID.t
```

Comment:
This would be a breaking change. Are we sure we want to remove this `Unknown` constructor? It might have been used for recovering some data when encountering unknown OIDs instead of failing completely.

Reply:
I agree in principle. But if I'm reading the code correctly, the only place that an `Algo.t` can go is into one of the functions `rsa_grammar`, `dsa_grammar`, `ec_grammar` or `dh_grammar`, and they will all generate a `parse_error "Algorithm OID and parameters don't match"` if they see an `Unknown`. I guess I could change the code in all of those functions to look something like this:

```
    let f = function
      | (Algo.Dsa, params) -> params
      | (Algo.Rsa, _) -> parse_error "Expecting DSA algorithm OID, got RSA"
      | (Algo.Dh, _) -> parse_error "Expecting DSA algorithm OID, got DH"
      | (Algo.Ec, _) -> parse_error "Expecting DSA algorithm OID, got EC"
      | (Algo.Unknown oid, _) -> parse_error "Unknown algorithm OID %a" Asn.OID.pp oid
```

To me this feels much more verbose, and like it's generating the error at the wrong place, but I guess it allows for other potential consumers of the `Algo` module to do different things with unknowns. What do you think?

Reply:
I do agree with Bertrand, IMHO we have to keep the `Unknown` constructor. 
But you're right we could also improve the error messages, which led to the current bugfix.

Why not keeping the constructor, and creating a `to_string` function for `Algo` module. This way, we'd solve the verbosity issue you highlighted and get more insights on the oid. Something like :
```
    let f = function
      | (Algo.Dsa, params) -> params
      | (toto, _) -> parse_error "Algorithm %s and parameters don't match" (Algo.to_string toto)
```